src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+
+std::size_t ResourceAwarePool::getPendingSize() const noexcept
+{
+  return m_pending_map.size();
+}
+
+std::size_t ResourceAwarePool::getActualSize() const noexcept
+{
+  return m_actual_bytes;
+}
+
+std::size_t ResourceAwarePool::getCurrentSize() const noexcept
+{
+  return m_current_bytes;
+}
+
+std::size_t ResourceAwarePool::getReleasableSize() const noexcept
+{
+  return m_releasable_bytes;
+}
+
+std::size_t ResourceAwarePool::getActualHighwaterMark() const noexcept
+{
+  return m_actual_highwatermark;
+}
+
+Platform ResourceAwarePool::getPlatform() noexcept
+{
+  return m_allocator->getPlatform();
+}
+
+camp::resources::Resource ResourceAwarePool::getResource(void* ptr) const
+{
+  for (auto& chunk : m_pending_map) { // check pending chunks
+    if (chunk->data == ptr) {
+      return chunk->m_resource;
+    }
+  }
+  auto it = m_used_map.find(ptr); // check used chunks
+  if (it != m_used_map.end()) {
+    auto chunk = it->second;
+    return chunk->m_resource;
+  }
+  for (auto pair = m_free_map.begin(); pair != m_free_map.end(); pair++) {
+    auto chunk = (*pair).second;
+    if (chunk->data == ptr) {
+      UMPIRE_LOG(
+          Warning,
+          fmt::format(
+              "Ptr {} corresponded to a free chunk in the ResourceAwarePool, so the resource may no longer be valid...",
+              ptr));
+      return chunk->m_resource;
+    }
+  }
+
+  UMPIRE_ERROR(runtime_error,
+               fmt::format("The pointer {} does not seem to be allocated with the ResourceAwarePool!", ptr));
+
+  return camp::resources::Host{}; // Function needs a return
+}
+
+MemoryResourceTraits ResourceAwarePool::getTraits() const noexcept
+{
+  return m_allocator->getTraits();
+}
+
+bool ResourceAwarePool::tracksMemoryUse() const noexcept
+{
+  return false;
+}
+
+std::size_t ResourceAwarePool::getBlocksInPool() const noexcept
+{
+  return m_used_map.size() + m_free_map.size() + m_pending_map.size();
+}
+
+std::size_t ResourceAwarePool::getLargestAvailableBlock() noexcept
+{
+  if (!m_free_map.size()) {
+    return 0;
+  }
+  return m_free_map.rbegin()->first;
+}
+
+void ResourceAwarePool::coalesce() noexcept
+{
+  UMPIRE_LOG(Debug, "()");
+
+  umpire::event::record([&](auto& event) {
+    event.name("coalesce").category(event::category::operation).tag("allocator_name", getName()).tag("replay", "true");
+  });
+
+  std::size_t suggested_size{m_should_coalesce(*this)};
+  if (0 != suggested_size) {
+    UMPIRE_LOG(Debug, "coalesce heuristic true, performing coalesce, suggested size is " << suggested_size);
+    do_coalesce(suggested_size);
+  }
+}
+
+void ResourceAwarePool::do_coalesce(std::size_t suggested_size) noexcept
+{
+  if (m_free_map.size() > 1) {
+    m_is_coalescing = true;
+    UMPIRE_LOG(Debug, "()");
+    release();
+    std::size_t size_post{getActualSize()};
+
+    if (size_post < suggested_size) {
+      std::size_t alloc_size{suggested_size - size_post};
+
+      camp::resources::Resource r = camp::resources::Host().get_default();
```

Comment:
Is hardcoding to the Host resource the right thing to do?

Reply:
Yes. David and I discussed this briefly, but essentially the coalesce will only ever happen on free chunks, so it doesn't really matter. Once they are reallocated the resource will be reset.

Reply:
Could you add a comment to that effect?

Reply:
Yes, Done!

Reply:
This is related to my "default resource" comment on the allocate(size) method.